File: src/writer.rs
use crate::optimizer::OptimizedRules;
use anyhow::{Context, Result};
use std::fs::File;
use std::io::{BufWriter, Write};
use std::path::Path;
use time::OffsetDateTime;

#[derive(Debug, Clone)]
pub struct ListSummary {
    pub name: String,
    pub ok: bool,
    pub bytes: Option<usize>,
    pub included_files: u64,
    pub total_lines: u64,
    pub network_rules: u64,
    pub cosmetic_rules: u64,
    pub empty: u64,
    pub unsupported: u64,
    pub invalid: u64,
    pub scriptlets_removed: u64,
    pub redirects_removed: u64,
    pub hosts_converted: u64,
    pub unsupported_options: u64,
    pub unsupported_cosmetic: u64,
    pub cosmetic_transforms: u64,
}

pub fn write_output(path: &Path, rules: &OptimizedRules, summaries: &[ListSummary]) -> Result<()> {
    let mut w =
        BufWriter::new(File::create(path).with_context(|| format!("creating {}", path.display()))?);

    let now = OffsetDateTime::now_utc();
    writeln!(w, "! StayBrave - optimized adblock-rust filter list")?;
    writeln!(
        w,
        "! Generated: {:04}-{:02}-{:02} {:02}:{:02}:{:02} UTC",
        now.year(),
        now.month() as u8,
        now.day(),
        now.hour(),
        now.minute(),
        now.second()
    )?;
    writeln!(w, "!")?;

    let fetched = summaries.iter().filter(|s| s.ok).count();
    writeln!(w, "! Sources ({fetched}/{} fetched):", summaries.len())?;
    for s in summaries {
        if s.ok {
            writeln!(
                w,
                "!   [ok]   {}: {} bytes (+{} included files), {} lines, {} network + {} cosmetic rules, {} empty, {} unsupported, {} invalid, {} hosts entries converted, {} scriptlets + {} redirects filtered, {} unsupported options, {} unsupported cosmetic rules, {} cosmetic rewrites",
                s.name,
                s.bytes.unwrap_or(0),
                s.included_files,
                s.total_lines,
                s.network_rules,
                s.cosmetic_rules,
                s.empty,
                s.unsupported,
                s.invalid,
                s.hosts_converted,
                s.scriptlets_removed,
                s.redirects_removed,
                s.unsupported_options,
                s.unsupported_cosmetic,
                s.cosmetic_transforms
            )?;
        } else {
            writeln!(w, "!   [fail] {}: fetch failed", s.name)?;
        }
    }
    writeln!(w, "!")?;

    let total_network: u64 = summaries.iter().map(|s| s.network_rules).sum();
    let total_cosmetic: u64 = summaries.iter().map(|s| s.cosmetic_rules).sum();
    let total_scriptlets: u64 = summaries.iter().map(|s| s.scriptlets_removed).sum();
    let total_redirects: u64 = summaries.iter().map(|s| s.redirects_removed).sum();
    let total_hosts: u64 = summaries.iter().map(|s| s.hosts_converted).sum();
    let total_uopts: u64 = summaries.iter().map(|s| s.unsupported_options).sum();
    let total_ucosm: u64 = summaries.iter().map(|s| s.unsupported_cosmetic).sum();
    let total_ctrans: u64 = summaries.iter().map(|s| s.cosmetic_transforms).sum();
    writeln!(
        w,
        "! Input rules: {} | Unique output: {} | Duplicates removed: {} | Cosmetic subsumed: {}",
        rules.input_rules, rules.unique_rules, rules.duplicates_removed, rules.cosmetic_subsumed
    )?;
    writeln!(
        w,
        "! Validated: {} network + {} cosmetic rules (adblock-rust parser)",
        total_network, total_cosmetic
    )?;
    writeln!(
        w,
        "! Filtered as unsupported: {} scriptlets + {} redirects",
        total_scriptlets, total_redirects
    )?;
    writeln!(
        w,
        "! Normalized: {} hosts entries converted | Eliminated: {} unsupported options, {} unsupported cosmetic rules | Rewritten: {} cosmetic rules",
        total_hosts, total_uopts, total_ucosm, total_ctrans
    )?;
    writeln!(w, "!")?;
    writeln!(
        w,
        "! Every rule below is validated by the adblock-rust engine parser."
    )?;
    writeln!(
        w,
        "! Unsupported uBO scriptlet injection and unlisted $redirect resources are removed."
    )?;
    writeln!(
        w,
        "! Procedural cosmetic rules are rewritten into forms the Brave engine executes."
    )?;

    for rule in &rules.rules {
        writeln!(w, "{rule}")?;
    }

    w.flush()?;
    Ok(())
}

File: src/writer.js
// Output writer: ABP-style header + per-source summary + final stats + rules.

import { writeFileSync, mkdirSync } from 'node:fs';
import { dirname } from 'node:path';

function pad(n, width = 2) {
  return String(n).padStart(width, '0');
}

export function timestamp(now = new Date()) {
  return [
    now.getUTCFullYear(),
    pad(now.getUTCMonth() + 1),
    pad(now.getUTCDate()),
    pad(now.getUTCHours()),
    pad(now.getUTCMinutes()),
  ].join('');
}

export function writeOutput(path, outputCfg, optimized, summaries) {
  const now = new Date();
  const chunks = [];
  chunks.push('[Adblock Plus 2.0]');
  chunks.push(`! Title: ${outputCfg.title}`);
  chunks.push(`! Version: ${timestamp(now)}`);
  chunks.push(`! Description: ${outputCfg.description}`);
  chunks.push(`! Expires: ${outputCfg.expires}`);
  chunks.push(`! Homepage: ${outputCfg.homepage}`);
  chunks.push(`! Last modified: ${now.toISOString().replace('T', ' ').slice(0, 19)} UTC`);
  chunks.push('!');
  chunks.push(`! Generated: ${now.toISOString().replace('T', ' ').slice(0, 19)} UTC`);
  chunks.push('!');

  const fetched = summaries.filter((s) => s.ok).length;
  chunks.push(`! Sources (${fetched}/${summaries.length} fetched):`);
  if (summaries.length === 0) {
    chunks.push('!   (none configured)');
  }
  for (const s of summaries) {
    if (s.ok) {
      const extra = [
        s.scriptlets,
        s.html_filters,
        s.responseheaders,
      ].join(' + ');
      chunks.push(
        `!   [ok]   ${s.name}: ${s.bytes} bytes (+${s.included_files} included files), ${s.total_lines} lines, ${s.network_rules} network + ${s.cosmetic_rules} cosmetic rules (incl. scriptlets ${extra}), ${s.empty} empty, ${s.unsupported} unsupported, ${s.invalid} invalid, ${s.hosts_converted} hosts entries converted, ${s.scriptlets_removed} scriptlets filtered, ${s.unsupported_options} unsupported options, ${s.trusted_source_dropped} trusted-source dropped, ${s.unsupported_cosmetic} unsupported cosmetic rules, ${s.cosmetic_transforms} rewrites`
      );
    } else {
      chunks.push(`!   [fail] ${s.name}: fetch failed (${s.error ?? 'unknown'})`);
    }
  }
  chunks.push('!');

  chunks.push(
    `! Input rules: ${optimized.input_rules} | Unique output: ${optimized.unique_rules} | Final output: ${optimized.rules.length} | Duplicates removed: ${optimized.duplicates_removed} | Cosmetic subsumed: ${optimized.cosmetic_selectors_subsumed} | Procedural subsumed: ${optimized.procedural_subsumed} | Network subsumed: ${optimized.network_subsumed}`
  );
  if (optimized.canonicalized_rules > 0) {
    chunks.push(
      `! Rewrites: ${optimized.canonicalized_rules} net-option spelling(s) canonicalized (uBO synonym map; duplicates collapsed)`
    );
  }
  chunks.push(
    `! Engine-filtered cosmetics (dead in stock uBO): ${optimized.cosmetic_engine_dropped ?? 0}`
  );
  if ((optimized.cosmetic_dead_candidates_count ?? 0) > 0) {
    chunks.push(
      `! Cosmetic dead-hide gate (engine-certified removals): ${optimized.cosmetic_dead_removed ?? 0} removed of ${optimized.cosmetic_dead_candidates_count} candidate(s) with their hide already withdrawn by an exception`
    );
  }
  if ((optimized.cosmetic_dead_exception_candidates_count ?? 0) > 0) {
    chunks.push(
      `! Cosmetic dead-exception gate (engine-certified removals): ${optimized.cosmetic_dead_exception_removed ?? 0} removed of ${optimized.cosmetic_dead_exception_candidates_count} candidate(s) whose selector no hide uses`
    );
  }
  chunks.push(
    `! Wildcard-TLD $domain rules (kept to avoid broadening): ${optimized.wildcard_domain_rules}`
  );
  if ((optimized.superset_candidates ?? 0) > 0) {
    chunks.push(
      `! Superset gate (engine-certified removals): ${optimized.superset_removed ?? 0} removed of ${optimized.superset_candidates} candidate(s) probed through uBO's own static network engine`
    );
  }
  if ((optimized.engine_dead_exception_candidates_count ?? 0) > 0) {
    chunks.push(
      `! Dead-exception gate (engine-certified removals): ${optimized.engine_dead_exception_removed ?? 0} removed of ${optimized.engine_dead_exception_candidates_count} candidate(s) that uBO proves suppress no surviving block`
    );
  }
  const tb = optimized.token_buckets ?? {};
  const tk = tb.tokened ?? optimized.hostname_tokened ?? 0;
  const jo = tb.justOrigin ?? optimized.just_origin ?? 0;
  const ca = tb.catchAll ?? optimized.catch_all_estimated ?? 0;
  chunks.push(
    `! Token buckets (SNFE-mirrored): ${tk} tokened, ${jo} just-origin, ${ca} catch-all network rules (last two are visited on every request)`
  );
  const eff = optimized.efficiency ?? {};
  if (eff.network !== undefined) {
    chunks.push(
      `! Efficiency (SNFE dispatch): network ${eff.network.grade} (${(eff.network.score * 100).toFixed(1)}%) | cosmetic ${eff.cosmetic.grade} (${(eff.cosmetic.score * 100).toFixed(1)}%)`
    );
  }
  chunks.push(
    `! Cosmetic channels: ${optimized.simple_class_id} simple class/id, ${optimized.complex_token_led} complex token-led, ${optimized.generic_misc} generic-misc, ${optimized.hostname_hide} hostname-hide, ${optimized.hostname_unhide} hostname-unhide, ${optimized.procedural} procedural`
  );
  chunks.push('!');
  chunks.push('! Built for Firefox uBO, where the exclusive capabilities live:');
  chunks.push(
    `!   ... CNAME uncloaking + $ipaddress= rules, ##^ HTML filters, ^responseheader filters, scriptlets`
  );
  const fe = optimized.firefox_exclusive ?? {};
  chunks.push(
    `! Firefox-exclusive rules shipped: ${fe.html_filters ?? 0} html_filters, ${fe.responseheaders ?? 0} responseheaders, ${fe.scriptlets ?? 0} scriptlets, ${fe.ipaddress ?? 0} ipaddress, ${fe.cname ?? 0} cname, ${fe.csp ?? 0} csp` +
      (fe.replace > 0 || fe.uritransform > 0 || fe.urlskip > 0
        ? ` | trusted-only: ${fe.replace ?? 0} replace, ${fe.uritransform ?? 0} uritransform, ${fe.urlskip ?? 0} urlskip`
        : '')
  );
  chunks.push('! Every rule below is validated by the uBlock Origin 1.74+ static-filter parser.');
  chunks.push(
    '! Unsupported uBO scriptlet injections, trusted-only ($replace=, $uritransform, $urlskip,'
  );
  chunks.push(
    '! trusted-* scriptlets) and dead syntax are removed unless keep_trusted_only is enabled;'
  );
  chunks.push(
    '! when enabled, add this list\'s URL to uBO\'s trustedListPrefixes advanced setting.'
  );
  chunks.push(
    '! Procedural cosmetic rules are rewritten into forms Firefox uBO executes'
  );
  chunks.push(
    '! (:contains -> :has-text, :nth-ancestor -> :upward, redundant :style stripped).'
  );
  chunks.push(
    '! Redundant network rules are subsumed by broader host/path rules;'
  );
  chunks.push(
    '! $badfilter pairs are stripped; $popup rules are preserved (uBO-native).'
  );
  chunks.push('');

  for (const rule of optimized.rules) {
    chunks.push(rule);
  }

  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, chunks.join('\n'));
  return rulesCount(optimized);
}

function rulesCount(o) {
  return o.rules.length;
}

export function formatSummaryLines(summaries) {
  const out = [];
  for (const s of summaries) {
    if (s.ok) {
      out.push(
        `${s.name}: ${s.network_rules} network, ${s.cosmetic_rules} cosmetic`
      );
    } else {
      out.push(`${s.name}: FAILED (${s.error})`);
    }
  }
  return out;
}